contracts/DefaultStrategy.sol

```diff
@@ -522,248 +436,288 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
      */
-    function getVotableGroups(uint256 votes) internal returns (address[] memory) {
-        uint256 numberGroups = activeGroups.length();
-        uint256 numberVotableGroups = 0;
-        address[] memory votableGroups = new address[](numberGroups);
+    function rebalance(address fromGroup, address toGroup) public {
+        if (!activeGroups.contains(fromGroup)) {
+            revert InvalidFromGroup(fromGroup);
+        }
 
-        for (uint256 i = 0; i < numberGroups; i++) {
-            address group = activeGroups.at(i);
-            uint256 scheduledVotes = account.scheduledVotesForGroup(group);
-            if (getElection().canReceiveVotes(group, votes + scheduledVotes)) {
-                votableGroups[numberVotableGroups] = group;
-                numberVotableGroups++;
-            }
+        if (!activeGroups.contains(toGroup)) {
+            revert InvalidToGroup(toGroup);
         }
 
-        address[] memory votableGroupsFinal = new address[](numberVotableGroups);
-        for (uint256 i = 0; i < numberVotableGroups; i++) {
-            votableGroupsFinal[i] = votableGroups[i];
+        (uint256 expectedFromStCelo, uint256 actualFromStCelo) = getExpectedAndActualStCeloForGroup(
+            fromGroup
+        );
+        if (actualFromStCelo <= expectedFromStCelo) {
+            // fromGroup needs to have more stCELO than it should
+            revert RebalanceNoExtraStCelo(fromGroup, actualFromStCelo, expectedFromStCelo);
+        }
+
+        (uint256 expectedToStCelo, uint256 actualToStCelo) = getExpectedAndActualStCeloForGroup(
+            toGroup
+        );
+
+        if (actualToStCelo >= expectedToStCelo) {
+            // toGroup needs to have less stCELO than it should
+            revert RebalanceEnoughStCelo(toGroup, actualToStCelo, expectedToStCelo);
         }
 
-        return votableGroupsFinal;
+        uint256 toMove = Math.min(
+            actualFromStCelo - expectedFromStCelo,
+            expectedToStCelo - actualToStCelo
+        );
+
+        updateGroupStCelo(fromGroup, toMove, false);
+        updateGroupStCelo(toGroup, toMove, true);
+        if (sorted) {
+            (address lesserKey, address greaterKey) = getLesserAndGreaterOfActiveGroupsWithdrawal(
+                fromGroup,
+                stCELOInGroup[fromGroup]
+            );
+            if (lesserKey != greaterKey) {
+                activeGroups.update(fromGroup, stCELOInGroup[fromGroup], lesserKey, greaterKey);
+                (lesserKey, greaterKey) = getLesserAndGreaterOfActiveGroupsDeposit(
+                    toGroup,
+                    stCELOInGroup[toGroup]
+                );
+                if (lesserKey != greaterKey) {
+                    activeGroups.update(toGroup, stCELOInGroup[toGroup], lesserKey, greaterKey);
+                    return;
+                }
+            }
+            sorted = false;
+        }
+        unsortedGroups.add(toGroup);
+        unsortedGroups.add(fromGroup);
```

Comment:
If I understand correctly, we mark both groups as unsorted even if we successfully sorted `fromGroup`? Could we avoid this by dealing with each group separately, and not nesting `if`s so deeply?

Could there be a `tryToSort(group)` helper that cleans up/deduplicates the above logic? 

Does the whole list need to be sorted for us to attempt sorting, or would it be enough for the given group we're trying to sort to not be in `unsortedGruops`?